Buildscript fixups and ClasspathLocator-related fixes

SizableShrimp opened the pull request:
This PR utilizes the new `inheritArgs` property in run configs that simplifies the `forge_test_data` run config.
ClasspathLocator's isValidManifest function has been fixed to use the right path from SJH, this fixes #7957. MOD_CLASSES is now reversed when parsed to deal with SJH currently only checking the last base path for a manifest. Since MOD_CLASSES always passes in with `resources;classes`, this is necessary so that manifests can be loaded in dev.

SizableShrimp committed:
Buildscript fixups, LIBRARY fix, and MOD_CLASSES fix

Closes #7957
Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

marchermans commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ClasspathLocator.java:
```diff
@@ -94,7 +94,7 @@ public void initArguments(Map<String, ?> arguments) {
 
     private static boolean isValidManifest(SecureJar sj) {
-        try (var jis = new JarInputStream(Files.newInputStream(sj.getRootPath()))) {
+        try (var jis = new JarInputStream(Files.newInputStream(sj.getPrimaryPath()))) {
```
This won't suffice for normal libraries.
The way CPL creates ModFiles requires a mods.toml to be available, so that they are loaded in the correct module.

See my other two PRs on this subject on how to fix that for the libraries.

SizableShrimp committed:
Fix CPL for LIBRARY jar types

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Load FMLModType libraries in the game layer

Made FMLCore a LANGPROVIDER so it loads in the plugin layer

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix ignore list to not use `forge-` anymore

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

LexManos commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/targets/CommonLaunchHandler.java:
```diff
@@ -72,6 +73,12 @@ record ExplodedModPath(String modid, Path path) {}
+        // The last path passed into SecureJar#from is the "main" path used to grab the manifest.
+        // ForgeGradle always passed in resources,classes with the expectation in 1.16 or lower that the resources would be checked first.
+        // So in 1.17+ FML/SJH, we need to reverse it.
+        modClassPaths.forEach((k, v) -> Collections.reverse(v));
```
So the problem isn't that things are in the wrong order its that things can quite often be in ANY order.
So the best option is the trust the order that exists, but put the one that specified the manifest files at the first one.
So instead of reverse. Sort.
Would a possible option be to copy over the manifest to every stream thats opened?
Or, since this is a read only file system. What we could do is make a fake input stream provider that sorts the entries/lists all files as if they were a single zip {merge together the zip headers}.

SizableShrimp replied:
This is removed/fixed as of [`6d9030b` (#7976)](https://github.com/MinecraftForge/MinecraftForge/pull/7976/commits/6d9030be92921e297c3859cf7145ed99f904fd4a) because SJH 0.9.51 and up properly checks ALL paths to find a manifest instead of just the last one.

SizableShrimp committed:
Fix Gradle wrapper for 7.2

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Bump SJH to 0.9.51

Reversing MOD_CLASSES is no longer necessary as 0.9.51 and up check all paths for a manifest

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add new `GAMELIBRARY` FML mod type for libraries on the game layer

Change FMLCore back to a library mod type
Update ForgeSPI to 4.0.10
Simplify manifest validation in ClasspathLocator
Pin more dependency versions

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Bump SJH to 0.9.52

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Only use ModJarMetadata if a jar has FMLModType MOD

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

LexManos commented:
As stated on discord, I am not a fan of you packaging gradle  spam and 'fixups' that arnt really fixups. It  makes reviewing your PRs a lot more annoying. This however is pending the fixes for the ignore list as discussed on discord.

LexManos commented on build.gradle:
```diff
@@ -392,7 +391,7 @@ def sharedFmlonlyForge = { Project prj ->
 
         // SecureJarHandler bootstrap values.
-        run.property 'ignoreList', prj.configurations.moduleonly.files.collect {it.name.replaceAll(/([-_]([.\d]*\d+)|\.jar$)/, '') }.join(',') + ",client-extra,fmlcore,javafmllanguage,mclanguage,${prj.name}-"
+        run.property 'ignoreList', prj.configurations.moduleonly.files.collect {it.name.replaceAll(/([-_]([.\d]*\d+)|\.jar$)/, '') }.join(',') + ",client-extra,fmlcore,javafmllanguage,mclanguage,${prj.name}-${prj.version},${prj.rootProject.ext.MC_VERSION}-${prj.name}${prj.version.substring(prj.rootProject.ext.MC_VERSION.length())}"
```
This is still a no.
Nothing in our environment is ever named MC_VERSION-forge-VERSION. So its just a complex filter for no reason.

LexManos commented:
Got tired of waiting on shrimp to finish this, so implemented a quick pass myself.
I do not have any test mods for this, so anyone who was running into this issue please report back with your results.
Note the introduction of GAMELIBRARY which adds things to the same classloader as mods {can access MC code} vs LIBRARY which adds them to the PLUGIN layer, which can't.
